src/qunit.cpp

```diff
@@ -2300,15 +2370,74 @@ void QUnit::CDIV(
     CMULx(&QInterface::CDIV, toMod, start, carryStart, length, controls, controlLen);
 }
 
-void QUnit::CMULModNOut(bitCapInt toMod, bitCapInt modN, bitLenInt inStart, bitLenInt outStart, bitLenInt length,
-    bitLenInt* controls, bitLenInt controlLen)
+void QUnit::CxMULModNOut(bitCapInt toMod, bitCapInt modN, bitLenInt inStart, bitLenInt outStart, bitLenInt length,
+    bitLenInt* controls, bitLenInt controlLen, bool inverse)
 {
-    if (controlLen == 0U) {
-        MULModNOut(toMod, modN, inStart, outStart, length);
+    // Try to optimize away the whole gate, or as many controls as is opportune.
+    std::vector<bitLenInt> controlVec;
+    if (CArithmeticOptimize(controls, controlLen, &controlVec)) {
+        // We've determined we can skip the entire operation:
         return;
     }
 
-    CMULModx(&QInterface::CMULModNOut, toMod, modN, inStart, outStart, length, controls, controlLen);
+    if (controlVec.size() == 0U) {
+        if (inverse) {
+            IMULModNOut(toMod, modN, inStart, outStart, length);
+        } else {
+            MULModNOut(toMod, modN, inStart, outStart, length);
+        }
+        return;
+    }
+
+    if (!inverse) {
+        SetReg(outStart, length, 0U);
+    }
+
+    // If "modN" is a power of 2, we have an optimized way of handling this.
+    if (pow2(log2(modN)) == modN) {
```

Comment:
I vastly prefer one of these:
```
int isPowerOfTwo (unsigned int x)
{
  return ((x != 0) && !(x & (x - 1)));
}
// OR
int isPowerOfTwo (unsigned int x)
{
  return ((x != 0) && ((x & (~x + 1)) == x));
}
```
(shamelessly stolen from https://www.exploringbinary.com/ten-ways-to-check-if-an-integer-is-a-power-of-two-in-c/)

Reply:
Replaced, (with source attribution).